File: 1325-52-2226-maximum-candies-allocated-to-k-children/1325-52-2226-maximum-candies-allocated-to-k-children.py
class Solution:
    def maximumCandies(self, candies: List[int], k: int) -> int:
        def ok(md):
            cnt = 0
            for x in candies:
                cnt+=x//md
            return cnt>=k
    
        st,ed,cur,md = 1,max(candies),0,0
        while st <=ed:
            md = (st+ed)//2
            if ok(md):
                cur = md
                st = md+1
            else:
                ed = md-1
        return cur

File: 1325-52-2226-maximum-candies-allocated-to-k-children/1325-52-2226-maximum-candies-allocated-to-k-children.java
class Solution {
    public int maximumCandies(int[] candies, long k) {
        int st = 1;
        int ed = max(candies);
        int cur = 0;

        while (st <= ed) {
            int md = (st + ed) / 2;
            if (ok(candies, md, k)) {
                cur = md;
                st = md + 1;
            } else {
                ed = md - 1;
            }
        }
        return cur;
    }

    private boolean ok(int[] candies, int md, long k) {
        long cnt = 0;
        for (int x : candies) {
            cnt += x / md;
        }
        return cnt >= k;
    }

    private int max(int[] arr) {
        int maxVal = Integer.MIN_VALUE;
        for (int num : arr) {
            if (num > maxVal) {
                maxVal = num;
            }
        }
        return maxVal;
    }

    
}